api/src/magic_share.rs

```diff
@@ -0,0 +1,263 @@
+        Some(s) => match s {
+            0 => v0_from_json(json),
+            _ => Err(anyhow!("Unsupported schema version")),
+        },
+        None => Err(anyhow!("Bad schema version")),
+    }
+}
+
+pub async fn find_by_id(db: &impl Database, id: &str) -> Result<MagicShare> {
+    for x in &[PREFIX_ONE_DAY, PREFIX_ONE_WEEK, PREFIX_ONE_MONTH] {
+        let key = format!("{}/{}", x, id);
+        match get_share(db, &key).await {
+            Ok(v) => { return Ok(v); },
+            Err(_) => ()
+        }
+    }
+
+    return Err(anyhow!("Not found"));
+}
+
+pub async fn get_share(
+    db: &impl Database,
+    location: &str,
+) -> Result<MagicShare> {
+    let bytes = db.get(location).await?;
+
+    let s: MagicShare = serde_json::from_slice(&bytes)?;
+    if Utc::now() > s.expires_at {
+        return Err(anyhow!("Expired"));
+    }
+
+    Ok(s)
+}
+
+pub async fn store_share(
+    db: &impl Database,
+    share: &MagicShare,
+) -> Result<String> {
+    match find_by_id(db, &share.id).await {
+        Ok(_) => return Err(anyhow!("Item already exists")),
+        _ => (),
+    };
+
+    let diff = share.expires_at - Utc::now();
+    let expiration_prefix = if diff < Duration::days(1) {
+        PREFIX_ONE_DAY
+    } else if diff < Duration::days(7) {
+        PREFIX_ONE_WEEK
+    } else {
+        PREFIX_ONE_MONTH
+    };
+
+    let key = format!("{}/{}", expiration_prefix, share.id);
+    let bytes = serde_json::to_vec(&share)?;
+
+    match db.put(&key, &bytes).await {
+        Ok(_) => Ok(key.to_string()),
+        Err(e) => Err(anyhow!(e)),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_new_v0_from_json() {
+        let bad_schema_version = "";
+        assert!(new_from_json(bad_schema_version.as_bytes()).is_err());
+
+        let unsupported_schema = "{\"schema_version\": 1000}";
+        assert!(new_from_json(unsupported_schema.as_bytes()).is_err());
+
+        let v = r#"
+        {
+            "schema_version": 0,
+            "id": "1234",
+            "expires_at": "2014-03-12T13:37:27+00:00",
+            "encrypted_credential": "some encrypted thing"
+        }"#;
+
+        let m = new_from_json(v.as_bytes()).unwrap();
+
+        assert_eq!(m.schema_version, 0);
+        //XXX: assert dates somehow...
+    }
+
+    #[cfg(not(feature = "s3"))]
+    use crate::store::FileStore;
+
+    #[cfg(not(feature = "s3"))]
+    #[async_std::test]
+    async fn test_share_roundtrip() {
+        let dir = tempfile::TempDir::new().unwrap();
+        let db = FileStore::new(dir.path(), "test", "v0").await.unwrap();
+
+        let m1 = MagicShare {
+            id: "1234".to_string(),
+            expires_at: Utc::now() + Duration::days(30),
+            schema_version: 0,
+            encrypted_credential: "5678".to_string(),
+        };
+
+        let location = store_share(&db, &m1).await.unwrap();
+        let r = get_share(&db, &location).await.unwrap();
+        assert_eq!(m1.id, r.id);
```

Comment:
I think you could derive partialeq for MagicShare and then just test equality instead of just the id. But not a big deal